File: src/NzbDrone.Core/MediaFiles/ChapterFile.cs
using System;
using NzbDrone.Common.Extensions;
using NzbDrone.Core.Datastore;
using NzbDrone.Core.MediaFiles.MediaInfo;

namespace NzbDrone.Core.MediaFiles
{
    // Sonarr divergence: NEW manga sibling per Phase 6 PIPELINE-04 — see DIVERGENCE.md.
    // Role-match analog: src/NzbDrone.Core/MediaFiles/EpisodeFile.cs (TV).
    // Imported chapter artifact (CBZ / folder of images) on disk + provenance fields.
    // Phase 16.1 D-06: 2 group-axis fields (TranslatedLanguage + ScanlationGroup), not 3 —
    // ReleaseGroup absorbed into ScanlationGroup (scanlation groups ARE the release groups
    // for manga). See DIVERGENCE.md ScanlationGroup-as-canonical entry.
    // Phase 8 cleanup: collapse with EpisodeFile when Tv/ deletes.
    public class ChapterFile : ModelBase
    {
        public int MangaId { get; set; }
        public int ChapterId { get; set; }
        public string RelativePath { get; set; }
        public string Path { get; set; }
        public long Size { get; set; }
        public DateTime DateAdded { get; set; }
        public string OriginalFilePath { get; set; }
        public string TranslatedLanguage { get; set; }   // BCP-47 — provenance

        // Sonarr divergence: ScanlationGroup is the canonical "release group" axis for manga;
        // renamed from Sonarr's EpisodeFile.ReleaseGroup because scanlation groups ARE the
        // release groups for manga. (D-06; DIVERGENCE.md ScanlationGroup-as-canonical entry.)
        public string ScanlationGroup { get; set; }

        // Phase 30 Plan 30-05 (II2-03) — ImageSharp probe result. Nullable: existing rows
        // pre-Migration-004 stay NULL (D-05 probe-on-import only — no backfill daemon).
        // Round-tripped via EmbeddedDocumentConverter<ChapterMediaInfo> Dapper TypeHandler
        // (registered in TableMapping.cs); TEXT column added by Migration 004.
        public ChapterMediaInfo MediaInfo { get; set; }

        public override string ToString()
        {
            return string.Format("[{0}] {1}", Id, RelativePath);
        }

        public string GetSceneOrFileName()
        {
            if (RelativePath.IsNotNullOrWhiteSpace())
            {
                return System.IO.Path.GetFileNameWithoutExtension(RelativePath);
            }

            if (Path.IsNotNullOrWhiteSpace())
            {
                return System.IO.Path.GetFileNameWithoutExtension(Path);
            }

            return string.Empty;
        }
    }
}

File: frontend/src/ChapterFile/ChapterFile.ts
// Sonarr divergence: NEW manga sibling per issue #84 (Plan 17.3-16 deferral
// resolution — Option A: author the peer dir to mirror frontend/src/Chapter/
// shape so the backend `src/NzbDrone.Core/MediaFiles/ChapterFile.cs` entity has
// a 1:1 frontend peer dir). See DIVERGENCE.md.
//
// Role-match analog: frontend/src/EpisodeFile/EpisodeFile.ts on the
// `v5-develop` Sonarr reference branch (shape-inspiration; the manga sibling
// trims TV-only fields per Phase 13 Plan 13-07 ChapterFileResource.cs).
//
// Manga sibling preserves: ModelBase extension + camelCase wire field set.
//
// Manga sibling diverges from EpisodeFile:
//   * Drop seasonNumber (PROJECT.md "Volumes/Seasons" Out-of-Scope).
//   * Drop quality (QualityModel), customFormats, customFormatScore,
//     qualityCutoffNotMet (Phase 5 D-05 — manga uses Custom Format +
//     Translation Profile, not QualityModel; the cutoff/CF columns belong to
//     MangaCutoffController + the Translation/CustomFormatProfile editors).
//   * Drop sceneName, indexerFlags, releaseType, mediaInfo (TV-only — no
//     scene-release naming, no torrent indexer flags, no video MediaInfo for
//     image-archive artifacts).
//   * Replace `seriesId` with `mangaId`; add `chapterId` (manga
//     ChapterFile is per-chapter, not per-(season,episode)).
//   * Replace `languages: Language[]` with `translatedLanguage?: string`
//     (BCP-47 single string per Phase 16.1 D-04 group-axis collapse).
//   * Rename `releaseGroup` to `scanlationGroup` (Phase 16.1 D-04/D-05/D-06 —
//     scanlation groups ARE the release groups for manga).
//
// Backend shape source: src/Mangarr.Api.V5/Manga/Chapter/ChapterFileResource.cs.
// Field set is exact; do NOT add fields the backend does not emit.
//
// Phase 8 cleanup: was — collapse with EpisodeFile when Tv/ deletes; Phase 17.3
// Plan 17.3-13 already retired the TV stub-dir family, so this directory is
// the canonical frontend home for chapter-file domain types/hooks/components.
import ModelBase from 'App/ModelBase';

interface ChapterFile extends ModelBase {
  mangaId: number;
  chapterId: number;
  relativePath?: string;
  path?: string;
  size: number;
  dateAdded: string;
  translatedLanguage?: string; // BCP-47; Phase 16.1 D-04
  scanlationGroup?: string; // Phase 16.1 D-04/D-05/D-06 — canonical release-group axis for manga
}

export default ChapterFile;
